File: Archive/JavaGraphics/src/com/graphics/input/Mouse.java
package com.graphics.input;

import org.lwjgl.glfw.GLFW;
import org.lwjgl.glfw.GLFWMouseButtonCallback;

import com.graphics.math.Vector3f;

public class Mouse extends GLFWMouseButtonCallback {

	public static boolean[] buttons = new boolean[10];
	public static Vector3f[] position = new Vector3f[10];
	@Override
	public void invoke(long window, int button, int action, int mods) {
		buttons[button] = action != GLFW.GLFW_RELEASE;
		position[button] = new Vector3f( MousePosition.getX(), MousePosition.getY(), 0.0f );
		System.out.println("Button " + button + " pressed at " + position[button].x + "," + position[button].y);
	}

	public static boolean isButtonDown( int button ) {
		return buttons[button];
	}

	public static Vector3f getPosition( int button ) {
		return position[button];
	}
}

File: Current/TheFinalCut/TheFinalCut/Graphics/Mouse.cpp
#include "Mouse.h"

double TFC::Graphics::Mouse::_x = 0.0;
double TFC::Graphics::Mouse::_y = 0.0;
bool TFC::Graphics::Mouse::_buttons[GLFW_MOUSE_BUTTON_LAST] = { 0 };
bool TFC::Graphics::Mouse::_buttonsDown[GLFW_MOUSE_BUTTON_LAST] = { 0 };
bool TFC::Graphics::Mouse::_buttonsUp[GLFW_MOUSE_BUTTON_LAST] = { 0 };

void TFC::Graphics::Mouse::MousePosCallback(GLFWwindow* window, double x, double y)
{
    int width, height;
    glfwGetFramebufferSize(window, &width, &height);
    _x = x;
    _y = height - y;
}

void TFC::Graphics::Mouse::MouseButtonCallback(GLFWwindow* window, int button, int action, int mods)
{
    if (button < 0)
        return;
    if (action != GLFW_RELEASE && _buttons[button] == false)
    {
        _buttonsDown[button] = true;
        _buttonsUp[button] = false;
    }

    if (action != GLFW_RELEASE && _buttons[button] == true)
    {
        _buttonsDown[button] = false;
        _buttonsUp[button] = true;
    }

    _buttons[button] = action != GLFW_RELEASE;
}

double TFC::Graphics::Mouse::GetX()
{
    return _x;
}

double TFC::Graphics::Mouse::GetY()
{
    return _y;
}

bool TFC::Graphics::Mouse::ButtonDown(int button)
{
    bool x = _buttonsDown[button];
    _buttonsDown[button] = false;
    return x;
}

bool TFC::Graphics::Mouse::ButtonUp(int button)
{
    bool x = _buttonsUp[button];
    _buttonsUp[button] = false;
    return x;
}

bool TFC::Graphics::Mouse::Button(int button)
{
    return _buttons[button];
}
